elastic/logstash
# geoip: extract database manager to stand-alone feature

## Release notes

Adds a general Geoip database manager that can be used by _any_ plugin that is running an Elastic-licensed disribution of Logstash.

## What does this PR do?

Introduces an Elastic-licensed GeoipDatabaseManagement tool that can be used by ANY plugin running on Elastic-licensed Logstash to retrieve a subscription to a GeoIP database that ensures EULA-compliance and frequent updates, and migrates the previous Elastic-licensed code-in-Logstash-core extension to the Geoip Filter to use this new tool, requiring ZERO changes to in-the-wild versions of the plugin.

The implementation of the new tool follows the previous implementation as closely as possible, but presents a new interface that ensures that a consumer can ATOMICALLY subscribe to a database path without risk that the subscriber will receive an update or expiry before it is finished applying the initial value:

~~~ ruby
geoip_manager = LogStash::GeoipDatabaseManagement::Manager.instance subscription = geoip_manager.subscribe('City')

subscription.observe(construct: ->(initial_dbinfo){ },
                     on_update: ->(updated_dbinfo){ },
                     on_expire: ->(       _      ){ })

subscription.release!
~~~

The GeoIP Filter's extension has new specs to validate that its previously
specified behavior holds true:

> ~~~
>     LogStash::Filters::Geoip
>       DatabaseManager
>         initialize
>           is not subscribed to the EULA manager
>         subscribe database path
>           when manager is disabled
>             when user subscribes to explicit path
>               logs about the path being configured manually
>               returns user input path
>               is not subscribed to the EULA manager
>             when user does not specify an explict path
>               logged about preparing CC
>               returns the CC-licensed database
>               is not subscribed to the EULA manager
>           when manager is enabled
>             when user does not specify an explicit path
>               and EULA database is pending
>                 logged about preparing CC
>                 logs about the user implicitly accepting the MaxMind EULA
>                 returns the CC-licensed database
>                 when the manager updates the db
>                   notifies the plugin
>                 when the manager expires the db
>                   notifies the plugin
>                 when the manager updates a different DB
>                   does not notify the plugin
>                 when the manager expires a different DB
>                   does not notify the plugin
>               and EULA database is expired
>                 logs about the user implicitly accepting the MaxMind EULA
>                 returns nil
>                 is subscribed for updates
>                 when the manager updates a different DB
>                   does not notify the plugin
>                 when the manager updates the db
>                   notifies the plugin
>               and EULA database has a recent database
>                 logs about the user implicitly accepting the MaxMind EULA
>                 returns the path to the managed database
>                 is subscribed for updates
>                 when the manager updates the db
>                   notifies the plugin
>                 when the manager expires the db
>                   notifies the plugin
>                 when the manager updates a different DB
>                   does not notify the plugin
>                 when the manager expires a different DB
>                   does not notify the plugin
>             when user subscribes to explicit path
>               logs about the path being configured manually
>               returns user input path
>               is not subscribed to the EULA manager
>         unsubscribe
>           removes plugin in state
>         shutdown
>           unsubscribes gracefully
> ~~~


## Why is it important/What is the impact to the user?

While the Geoip filter has long had an extension inside Logstash core to add a GeoIP database manager for itself when running in an Elastic-licensed Logstash distro, this feature makes a comparable Geoip database manager available to _any_ plugin, without the specific implementation details of the Geoip filter.

## Checklist


## Author's Checklist

- [x] Self-review commentary

## How to test this PR locally

Run pipelines that include the Geoip filter as set to use a managed database, observing that databases are refreshed as expected and maintained across process restarts

Closes #15387
